fri/src/verifier.rs

```diff
@@ -0,0 +1,282 @@
+use std::error::Error;
+
+use ark_ff::{FftField, PrimeField};
+use ark_poly::{domain::EvaluationDomain, Radix2EvaluationDomain};
+use randomness::Prng;
+use sha3::Digest;
+
+use crate::{lde::MultiplicativeLDE, parameters::FriParameters};
+use channel::{fs_verifier_channel::FSVerifierChannel, Channel, VerifierChannel};
+use commitment_scheme::{
+    make_commitment_scheme_verifier, table_verifier::TableVerifier, CommitmentHashes,
+};
+
+#[allow(dead_code)]
+pub type FirstLayerQueriesCallback<F> = fn(&[u64]) -> Vec<F>;
+
+#[allow(dead_code)]
+pub struct FriVerifier<
+    F: FftField + PrimeField,
+    P: Prng + Clone + 'static,
+    W: Digest + Clone + 'static,
+> {
+    channel: FSVerifierChannel<F, P, W>,
+    params: FriParameters<F, Radix2EvaluationDomain<F>>,
+    commitment_hashes: CommitmentHashes,
+    first_layer_callback: FirstLayerQueriesCallback<F>,
+    n_layers: usize,
+    first_eval_point: Option<F>,
+    eval_points: Vec<F>,
+    table_verifiers: Vec<TableVerifier<F, P, W>>,
+    query_indices: Vec<u64>,
+    query_results: Vec<F>,
+    expected_last_layer: Option<Vec<F>>,
+    eval_points_test: Option<Vec<F>>,
+    last_layer_coefficients_test: Option<Vec<F>>,
+}
+
+#[allow(dead_code)]
+impl<F: FftField + PrimeField, P: Prng + Clone + 'static, W: Digest + Clone + 'static>
+    FriVerifier<F, P, W>
+{
+    pub fn new(
+        channel: FSVerifierChannel<F, P, W>,
+        params: FriParameters<F, Radix2EvaluationDomain<F>>,
+        commitment_hashes: CommitmentHashes,
+        first_layer_callback: FirstLayerQueriesCallback<F>,
+        eval_points_test: Option<Vec<F>>,
+        last_layer_coefficients_test: Option<Vec<F>>,
+    ) -> Self {
+        let n_layers = params.fri_step_list.len();
+        Self {
+            channel,
+            params,
+            commitment_hashes,
+            first_layer_callback,
+            n_layers,
+            first_eval_point: None,
+            eval_points: vec![F::zero(); n_layers - 1],
+            table_verifiers: vec![],
+            query_indices: vec![],
+            query_results: vec![],
+            expected_last_layer: None,
+            eval_points_test,
+            last_layer_coefficients_test,
+        }
+    }
+
+    pub fn verify_fri(&mut self) -> Result<(), Box<dyn Error>> {
```

Comment:
Convert the error instead: handling a `dyn Error` seems like a pain.

Using anyhow is fine :)